File: src/main/java/me/yellowbear/uwujobs/UwuJobs.java
package me.yellowbear.uwujobs;

import co.aikar.commands.PaperCommandManager;
import co.aikar.idb.*;
import me.yellowbear.uwujobs.commands.JobsCommand;
import me.yellowbear.uwujobs.jobs.BlockBreak;
import me.yellowbear.uwujobs.jobs.BlockPlace;
import me.yellowbear.uwujobs.jobs.MobKill;
import me.yellowbear.uwujobs.services.ConfigService;
import org.bukkit.command.CommandExecutor;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.IOException;
import java.sql.SQLException;
import java.util.HashSet;
import java.util.Set;

public final class UwuJobs extends JavaPlugin implements Listener, CommandExecutor {
    private BlockSets blockSets = new BlockSets();
    @Override
    public void onEnable() {
        try {
            ConfigService.registerCustomConfig("blocks.yml");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        ConfigService.registerService(blockSets, "blocks.yml");
        ConfigService.loadConfigs();
        try {
            getServer().getPluginManager().registerEvents(new UwuJobs(), this);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        // Setup ACF
        PaperCommandManager manager = new PaperCommandManager(this);

        manager.getCommandCompletions().registerCompletion("jobs", c -> {
            Set<String> jobs = new HashSet<>();
            for (BlockBreak job : BlockBreak.values()) {
                jobs.add(job.name().toLowerCase());
            }
            return jobs;
        });

        manager.registerCommand(new JobsCommand());

        // Setup database
        DatabaseOptions options = DatabaseOptions.builder().sqlite("plugins/uwuJobs/uwu.db").build();
        Database db = PooledDatabaseOptions.builder().options(options).createHikariDatabase();
        DB.setGlobalDatabase(db);

        try {
            for (BlockBreak job : BlockBreak.values()) {
                DB.executeInsert(String.format("CREATE TABLE IF NOT EXISTS %s (id TEXT UNIQUE, xp INT, next INT)", job.name().toLowerCase()));
            }
            for (BlockPlace job : BlockPlace.values()) {
                DB.executeInsert(String.format("CREATE TABLE IF NOT EXISTS %s (id TEXT UNIQUE, xp INT, next INT)", job.name().toLowerCase()));
            }
            for (MobKill job : MobKill.values()) {
                DB.executeInsert(String.format("CREATE TABLE IF NOT EXISTS %s (id TEXT UNIQUE, xp INT, next INT)", job.name().toLowerCase()));
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void onDisable() {
        DB.close();
        // Plugin shutdown logic
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        for (BlockBreak job : BlockBreak.values()) {
            try {
                DB.executeInsert(String.format("INSERT INTO %s (id, xp) VALUES ('%s', %s)", job.name().toLowerCase(), event.getPlayer().getUniqueId(), 1));
            } catch (SQLException e) {
                // player already exists
            }
        }
        for (BlockPlace job : BlockPlace.values()) {
            try {
                DB.executeInsert(String.format("INSERT INTO %s (id, xp) VALUES ('%s', %s)", job.name().toLowerCase(), event.getPlayer().getUniqueId(), 1));
            } catch (SQLException e) {
                // player already exists
            }
        }
        for (MobKill job : MobKill.values()) {
            try {
                DB.executeInsert(String.format("INSERT INTO %s (id, xp) VALUES ('%s', %s)", job.name().toLowerCase(), event.getPlayer().getUniqueId(), 1));
            } catch (SQLException e) {
                // player already exists
            }
        }

        /*String[] board = new String[] {
          "my",
          "awesome" ,
          "scoreboard",
          ":3"
        };
        ScoreboardController controller = new ScoreboardController(event.getPlayer(), board);*/
    }

    @EventHandler
    public void onBlockMined(BlockBreakEvent event) throws IOException {
        Job.handleJobEvent(event, BlockSets.breakJobsMap);
    }
    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event) throws IOException {
        Job.handleJobEvent(event, BlockSets.placeJobsMap);
    }
    @EventHandler
    public void onEntityDeath(EntityDeathEvent event) throws IOException {
        Job.handleJobEvent(event, BlockSets.killJobsMap);
    }
}

File: src/main/java/me/yellowbear/uwujobs/UwuJobs.kt
package me.yellowbear.uwujobs

import co.aikar.commands.PaperCommandManager
import me.yellowbear.uwujobs.commands.Jobs
import me.yellowbear.uwujobs.jobs.JobPlayer
import org.bukkit.command.CommandExecutor
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockFertilizeEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.plugin.java.JavaPlugin
import java.sql.SQLException
import java.util.concurrent.TimeUnit
import me.yellowbear.uwujobs.jobs.Jobs.Companion.jobs as jobsList

class UwuJobs : JavaPlugin(), Listener, CommandExecutor {
    override fun onEnable() {
        Config.loadJobs()
        Config.loadConfig()

        logger.info(jobsList.toString())

        try {
            server.pluginManager.registerEvents(UwuJobs(), this)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        // Setup ACF
        val manager = PaperCommandManager(this)

        manager.registerCommand(Jobs())

        manager.commandCompletions.registerCompletion("jobs") {
            val jobs: MutableSet<String> = HashSet()
            for (job in Config.jobs) {
                jobs.add(job.name.lowercase())
            }
            jobs.add("all")
            jobs
        }

        // Setup database
        Database.connect()

        try {
            val statement = Database.dataSource.connection.createStatement()
            for (job in Config.jobs) {
                statement.execute("CREATE TABLE IF NOT EXISTS ${job.name.lowercase()} (id VARCHAR(36) PRIMARY KEY, xp INT)")
            }
            statement.close()
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }

        server.asyncScheduler.runAtFixedRate(this, {
            if (server.onlinePlayers.isEmpty()) return@runAtFixedRate

            try {
                val connection = Database.dataSource.connection
                val statement = connection.createStatement()
                for (job in jobsList) {
                    for (player in job.value) {
                        player.value.saveToDb(statement, player.key, job.key)
                    }
                }
                statement.close()
                connection.close()
                logger.info("Saved to database")
            } catch (e: SQLException) {
                throw RuntimeException(e)
            }
        }, Config.config.save_interval, Config.config.save_interval, TimeUnit.SECONDS)
    }

    override fun onDisable() {
        Database.dataSource.close()
        // Plugin shutdown logic
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        try {
            val statement = Database.dataSource.connection.createStatement()
            for (job in Config.jobs) {
                val row =
                    statement.executeQuery("SELECT xp FROM ${job.name.lowercase()} WHERE id = '${event.player.uniqueId}'")
                if (!row.next()) {
                    statement.execute("INSERT INTO ${job.name.lowercase()} (id, xp) VALUES ('${event.player.uniqueId}', 0)")
                    jobsList[job.name.lowercase()]?.set(
                        event.player.uniqueId.toString(),
                        JobPlayer(0)
                    )
                } else {
                    jobsList[job.name.lowercase()]?.set(
                        event.player.uniqueId.toString(),
                        JobPlayer(row.getInt("xp"))
                    )
                }

            }
            statement.close()
        } catch (e: SQLException) {
            throw RuntimeException(e)
        }
    }


    @EventHandler
    fun onBlockBreak(event: BlockBreakEvent) {
        val player = event.player
        //TODO: Check for crop age

        for (job in Config.jobs) {
            for (reward in job.rewards) {
                if (reward.brokenBlocks == null) continue
                for (block in reward.brokenBlocks) {
                    if (block == event.block.type.name) {
                        val jobPlayer: JobPlayer? = jobsList[job.name.lowercase()]!![player.uniqueId.toString()]
                        if (jobPlayer == null) {
                            logger.warning("Player ${player.name} has no jobPlayer for job ${job.name}")
                            return
                        }
                        Level.awardXp(event.player, jobPlayer, reward.amount, job)
                    }
                }
            }
        }

    }

    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent) {
        val player = event.player

        for (job in Config.jobs) {
            for (reward in job.rewards) {
                if (reward.placedBlocks == null) continue
                for (block in reward.placedBlocks) {
                    if (block == event.block.type.name) {
                        val jobPlayer: JobPlayer? = jobsList[job.name.lowercase()]!![player.uniqueId.toString()]
                        if (jobPlayer == null) {
                            logger.warning("Player ${player.name} has no jobPlayer for job ${job.name}")
                            return
                        }
                        Level.awardXp(event.player, jobPlayer, reward.amount, job)
                    }
                }
            }
        }
    }

    @EventHandler
    fun onEntityDeath(event: EntityDeathEvent) {
        val player = event.entity.killer
        if (player == null) return

        for (job in Config.jobs) {
            for (reward in job.rewards) {
                if (reward.killedEntities == null) continue
                for (entity in reward.killedEntities) {
                    if (entity == event.entity.type.name) {
                        val jobPlayer: JobPlayer? = jobsList[job.name.lowercase()]!![player.uniqueId.toString()]
                        if (jobPlayer == null) {
                            logger.warning("Player ${event.entity.killer} has no jobPlayer for job ${job.name}")
                            return
                        }
                        Level.awardXp(player, jobPlayer, reward.amount, job)
                    }
                }
            }
        }
    }

    @EventHandler
    fun onFertilize(event: BlockFertilizeEvent) {
        val player = event.player
        if (player == null) return

        for (job in Config.jobs) {
            for (reward in job.rewards) {
                if (reward.fertilizedBlocks == null) continue
                for (block in reward.fertilizedBlocks) {
                    if (block == event.block.type.name) {
                        val jobPlayer: JobPlayer? = jobsList[job.name.lowercase()]!![player.uniqueId.toString()]
                        if (jobPlayer == null) {
                            logger.warning("Player ${player.name} has no jobPlayer for job ${job.name}")
                            return
                        }
                        Level.awardXp(player, jobPlayer, reward.amount, job)
                    }
                }
            }
        }
    }
}
